src/handlers/events.rs

```diff
@@ -29,22 +28,55 @@ pub async fn get_event_buffer(
     Extension(state): Extension<AppState>,
     AuthBearer(token): AuthBearer,
 ) -> Result<Json<Vec<Event>>, Error> {
-    let mut ret = Vec::new();
-    for event in state.events_buffer.lock().await.get_ref().iter().rev() {
-        let requester = try_auth(&token, state.users.lock().await.get_ref()).ok_or(Error {
-            inner: ErrorInner::PermissionDenied,
-            detail: "Token error".to_string(),
-        })?;
-        if can_user_view_event(event, &requester) {
-            ret.push(event.clone());
-        }
-    }
-    Ok(Json(ret))
+    let requester = try_auth(&token, state.users.lock().await.get_ref()).ok_or(Error {
+        inner: ErrorInner::PermissionDenied,
+        detail: "Token error".to_string(),
+    })?;
+    Ok(Json(
+        state
+            .events_buffer
+            .lock()
+            .await
+            .get_ref()
+            .iter()
+            .rev()
+            .filter(|event| can_user_view_event(event, &requester))
+            .cloned()
+            .collect(),
+    ))
+}
+
+pub async fn get_console_out_buffer(
+    Extension(state): Extension<AppState>,
+    AuthBearer(token): AuthBearer,
+    Path(uuid): Path<String>,
+) -> Result<Json<Vec<Event>>, Error> {
+    let requester = try_auth(&token, state.users.lock().await.get_ref()).ok_or(Error {
+        inner: ErrorInner::PermissionDenied,
+        detail: "Token error".to_string(),
+    })?;
+    Ok(Json(
+        state
+            .console_out_buffer
+            .lock()
+            .await
+            .get_ref()
+            .iter()
+            .rev()
+            .filter(|event| {
+                event.instance_uuid == uuid
+                    && matches!(event.event_inner, EventInner::InstanceOutput(_))
+                    && can_user_view_event(event, &requester)
+            })
+            .cloned()
+            .collect(),
+    ))
 }
 
 #[derive(Deserialize)]
 pub struct WebsocketQuery {
     token: String,
+    uuid: String,
```

Comment:
the overall "event_stream" also use this struct, but doesn't need the UUID
make them 2 structs instead?

Reply:
UUID is now required, and an uuid of "all" sends back all the events

Reply:
"all" feels sketchy but ok